EbThreads: Replace thread create function with the one from SVT-AV1

wader opened the pull request:
Current code will erroneously try to cancel thread when failing to create it causing a segfault.

Now uses same scheduling parameter as SVT-AV1.

wader commented:
Some investigation and repro steps leading to this in https://github.com/wader/static-ffmpeg/issues/301

wader commented:
BTW i failed to configure a container with capabilities and permission to allow creating a thread with scheduling config, maybe there is something fishy with the config?

wader commented on Source/Lib/Codec/EbThreads.c:
```diff
@@ -84,7 +84,6 @@ EbHandle eb_vp9_create_thread(
-                pthread_cancel(*((pthread_t*)thread_handle));
                 free(thread_handle);
 
                 thread_handle = (pthread_t*)malloc(sizeof(pthread_t));
```
unrelated: the above free and malloc could be skipped i think? but maybe good for clarify

1480c1 commented:
you may instead want to adapt what has been changed in SVT-AV1's code https://gitlab.com/AOMediaCodec/SVT-AV1/-/blob/master/Source/Lib/Common/Codec/EbThreads.c#L66 rather than what we have in svt-vp9. Although you could probably skip the `Failed to set thread priority` message since it's spammy inside containers

wader commented:
Ah ok, hmm so just copy EbThreads.* from SVT-AV1 should just work™️? seems like the av1 one sets schedule parameters a bit differently, have no idea how important the current config is?

1480c1 commented:
I don't think it's very important if we dropped it for svt-av1

wader commented:
Now updated to copy code from `svt_create_thread` from SVT-AV1, changed code to use SVT_LOG and to not log `pthread_setschedparam` failure.

Have only tested in a alpine 3.17 container.

1480c1 commented:
I seem to get a few new warnings with this PR
```c
/code/SVT-VP9/Source/Lib/Codec/EbThreads.c: In function ‘eb_vp9_create_thread’:
/code/SVT-VP9/Source/Lib/Codec/EbThreads.c:87:50: warning: implicit declaration of function ‘strerror’; did you mean ‘perror’? [-Wimplicit-function-declaration]
   87 |         SVT_LOG("Failed to create thread: %s\n", strerror(errno));
      |                                                  ^~~~~~~~
      |                                                  perror
/code/SVT-VP9/Source/Lib/Codec/EbThreads.c:87:44: warning: format ‘%s’ expects argument of type ‘char *’, but argument 2 has type ‘int’ [-Wformat=]
   87 |         SVT_LOG("Failed to create thread: %s\n", strerror(errno));
      |                                           ~^     ~~~~~~~~~~~~~~~
      |                                            |     |
      |                                            |     int
      |                                            char *
      |                                           %d
/code/SVT-VP9/Source/Lib/Codec/EbThreads.c:106:10: warning: implicit declaration of function ‘geteuid’ [-Wimplicit-function-declaration]
  106 |     if (!geteuid()) {
      |          ^~~~~~~
```
I believe if you include the proper headers for the two functions, it should resolve the warnings
```c
#include <string.h>
#include <unistd.h>
```

wader commented:
@1480c1 ah sorry just copy pasted, should have checked. fixed now.

1480c1 commented:
Compiled fine locally now, thanks

wader commented on Source/Lib/Codec/EbThreads.c:
```diff
@@ -45,60 +47,69 @@ void printfTime(const char *fmt, ...)
+     * We still need to exclude the case of thread sanitizer because we
+     * run the test as root inside the container and trying to change
+     * the thread priority will __always__ fail the thread sanitizer.
+     * https://github.com/google/sanitizers/issues/1088
```
I removed some !SANITIZER from the if below, maybe this comment should be removed also?

1480c1 replied:
Well, the comment is probably fine for anyone who might be interested enough to try to run svt-vp9 under a sanitizer

wader replied:
👍
